### ISSUE TITLE
IPVS low throughput

### ISSUE THREAD

**Issue by annProg** [kubernetes/kubernetes#70747]
**What happened**:  
Service `A` with clusterIP `169.169.14.147` and a domain `k8s.dev.example.com`, the domain resolve to ingress-nginx. Anothor service `B` will request `A`, when use domain it works fine. I think using clusterIP directly will get better performance because it saves some time for dns resolve and ingress-nginx proxy, so I change domain to clusterIP. But the result was not expected, it got worse, very low throughput and high latency.

**What you expected to happen**:
On the above situation, clusterIP should have better performance than ingress-nginx.

**How to reproduce it (as minimally and precisely as possible)**:
`ipvs` mode, do test with ab
```
# clusterIP
ab -c100 -t300 -n20000000 -r http://169.169.14.147/

# domain resolve to ingress-nginx
ab -c100 -t300 -n20000000 -r http://k8s.dev.example.com/
```

**Anything else we need to know?**:
I install an ipvs enviroment manually and do some test, and found it possiblly because ipvs will use iptables to realize `SNAT`

![ipvs](https://user-images.githubusercontent.com/4145852/48137324-c1fbfe80-e2dc-11e8-9fba-ffce3260186d.png)

**Environment**:
- Kubernetes version (use `kubectl version`):  1.11.1
- Cloud provider or hardware configuration: 
- OS (e.g. from /etc/os-release):  centos 7.2
- Kernel (e.g. `uname -a`): 3.10
- Install tools: binary
- Others:

/kind bug

**Comment by annProg** [comment]
@kubernetes/sig-network-bugs

**Comment by annProg** [comment]
I try to change some kernel param and found `conn_reuse_mode` is related to this issue
`net.ipv4.vs.conn_reuse_mode` , default `1`. change to `0`，then complte about 3000k request, pretty good performance.

So, should I change the param in production?  Does it have any side effects？

**Comment by annProg** [comment]
@m1093782566 Can you look at it?

**Comment by m1093782566** [comment]
> I try to change some kernel param and found conn_reuse_mode is related to this issue
net.ipv4.vs.conn_reuse_mode , default 1. change to 0，then complte about 3000k request, pretty good performance.

Yes, seems you find the key to the problem - it's set to 0 in my environment and I have not found any issue so far(it's about IPVS module behavior and we will not find net.ipv4.vs.conn_reuse_mode parameter in old kernels ).

BTW, you can also set the ipvs conntrack(https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L163) to 0 since we are no longer depend on conntrack any more.

**Comment by annProg** [comment]
> > I try to change some kernel param and found conn_reuse_mode is related to this issue
> > net.ipv4.vs.conn_reuse_mode , default 1. change to 0，then complte about 3000k request, pretty good performance.
> 
> Yes, seems you find the key to the problem - it's set to 0 in my environment and I have not found any issue so far(it's about IPVS module behavior and we will not find net.ipv4.vs.conn_reuse_mode parameter in old kernels ).
> 
> BTW, you can also set the ipvs conntrack(https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L163) to 0 since we are no longer depend on conntrack any more - it will improve performance as well.

Thanks for reply.

I am using kubernetes `1.11.1` , and after set `sysctl -w net.ipv4.vs.conntrack=0`, clusterIP can not be reachable.  Which version do I need ?

**Comment by annProg** [comment]
> Sorry, Cluster IP still need the IPVS conntrack and nodeport does not rely on it any more - we are trying to figure it out for Cluster IP.

Got it, thanks!

**Comment by m1093782566** [comment]
cc @Lion-Wei 

Should we tune the kernel parameter in kube-proxy?

**Comment by dawidmalina** [comment]
There is a nice article describing even more kernel parameters to tune ipvs: https://jsravn.com/2017/12/24/ipvs-with-kubernetes-ingress/

**Comment by annProg** [comment]
Side effect  like this comment: https://github.com/cloudnativelabs/kube-router/issues/544#issuecomment-431553463

<img width="790" alt="hpa_prize" src="https://user-images.githubusercontent.com/4145852/48302647-118b3600-e53a-11e8-9ac6-ff96c66f8e82.png">

Note: time in picture below is UTC+8. 
![hpa_price](https://user-images.githubusercontent.com/4145852/48302771-d0942100-e53b-11e8-9edb-bc4097cae476.png)


Abviously every time scaledown , response time will increase.
 @m1093782566  Any idea for this?

**Comment by annProg** [comment]
@m1093782566  yes, I had test these parameters
- `net.ipv4.vs.sloppy_tcp` not found in centos 7.2(kernel 3.10)
- `net.ipv4.vs.expire_nodest_conn`  will not affect the test
- `net.ipv4.vs.conn_reuse_mode=2` is similar with `1`, will be low throughput

According to this issue https://github.com/kubernetes/kubernetes/issues/57841 ,I guess it may be the root cause, so I upgrade my setup from 1.11.1 to 1.12.2. 
But did not solve this problem, and found a new problem, rs delete failed after scaledown. I had open a new issue to this problem: https://github.com/kubernetes/kubernetes/issues/70997

**Comment by Lion-Wei** [comment]
Hi, @annProg @dawidmalina @m1093782566 , I raised a pr for this issue. Please check. : )
@jsravn Looking forward for your advise.

**Comment by m1093782566** [comment]
@annProg 

Could you please check out the HEAD to see if your issues have been resolved?

**Comment by jsravn** [comment]
It's not a big surprise this benchmark performs better with `conn_reuse_mode=0`. `ab` uses http/1.0 so is creating a new connection for every single request. With `conn_reuse_mode=0` IPVS will reuse the existing connection entry - sending repeated requests to the same real server rather than actually load balancing. So my guess is you will see a very unbalanced load with using `conn_reuse_mode=0`, but it makes the benchmark look better.

I'm not sure this benchmark is a realistic scenario. I suggest using a tool like `wrk` instead which uses persistent connections and see what happens.

**Comment by jsravn** [comment]
Apart from the load balancing issue, it's also possible the overhead of `conn_reuse_mode` is bad when creating `20000000` connections in a short period of time. But I'm not sure we should be optimizing for such a thing, when in reality connections will be much longer lived. It's a trade off between improved load balancing (conn_reuse_mode=1 or 2) and connection creation latency (conn_reuse_mode=0).

**Comment by annProg** [comment]
@jsravn I have a php program use unirest to request  other sevice in the cluster. It not use keepalive.

**Comment by m1093782566** [comment]
I would prefer to improve the latency number.

**Comment by Lion-Wei** [comment]
> Abviously every time scaledown , response time will increase.
@annProg , I think this was because the the container already deleted, but kube-proxy need time to flush the rules, iptables mode should also have this problem.

**Comment by juliantaylor** [comment]
Here the likely reason for the slow throughput is described:
https://marc.info/?l=linux-virtual-server&m=151706660530133&w=2

As disabling `conn_reused_mode` causes https://github.com/kubernetes/kubernetes/issues/81775, is any of the options listed possible to do?

**Comment by yyx** [comment]
Hello everyone：
      We are very fortunate to tell you that this bug has been fixed by us and has been verified to work very well.  The patch（[ipvs: avoid drop first packet to reuse conntrack](https://lkml.org/lkml/2020/6/11/272)） is being submitted to the Linux kernel community. You can also apply this patch to your own kernel, and then only need to set net.ipv4.vs.conn_reuse_mode=1. 
That Can solve the following problems:
1. Rolling update, IPVS keeps scheduling traffic to the destroyed Pod
2. Unbalanced IPVS traffic scheduling after scaled up or rolling update
3.fix IPVS low throughput issue #71114
https://github.com/kubernetes/kubernetes/pull/71114
4.One second connection delay in masque
https://marc.info/?t=151683118100004&r=1&w=2
5.IPVS low throughputÂ #70747
https://github.com/kubernetes/kubernetes/issues/70747  
6. Apache Bench can fill up ipvs service proxy in seconds #544
https://github.com/cloudnativelabs/kube-router/issues/544
7. Additional 1s latency in `host -> service IP -> pod` when upgrading from `1.15.3 -> 1.18.1` on RHEL 8.1 #90854
https://github.com/kubernetes/kubernetes/issues/90854

  Thank you.
  By Yang Yuxi （TencentCloudContainerTeam）

**Comment by andrewsykim** [comment]
Following-up on @yyx's comment above for posterity. 

The patch mentioned above didn't make it to the kernel but there are two recently merged patches worth highlighting. One of them fixes the 1 second delay issue and the other fixes dropped packets when stale connection entries in the IPVS table are used:
1) http://patchwork.ozlabs.org/project/netfilter-devel/patch/20200701151719.4751-1-ja@ssi.bg/ 
2) http://patchwork.ozlabs.org/project/netfilter-devel/patch/20200708161638.13584-1-kim.andrewsy@gmail.com/
